domain/src/main/java/org/oppia/domain/topic/StoryProgressController.kt

```diff
@@ -32,27 +42,229 @@ private val FRACTIONS_COMPLETED_CHAPTERS = listOf(FRACTIONS_EXPLORATION_ID_0)
+
+  /** Indicates that the given story id does not have any associated story progress. */
+  class StoryProgressNotFoundException(msg: String) : java.lang.Exception(msg)
+
+  /** Indicates that the given topic id does not have any associated topic progress. */
+  class TopicProgressNotFoundException(msg: String) : java.lang.Exception(msg)
+
+  /**
+   * These Statuses correspond to the exceptions above such that if the deferred contains
+   * EXPLORATION_NOT_FOUND, the [ExplorationNotFoundException] will be passed to a failed AsyncResult.
+   *
+   * SUCCESS corresponds to a successful AsyncResult.
+   */
+  private enum class StoryProgressActionStatus {
+    SUCCESS,
+    EXPLORATION_ALREADY_COMPLETED,
+    EXPLORATION_NOT_FOUND,
+    STORY_NOT_FOUND,
+    TOPIC_NOT_FOUND,
+    STORY_PROGRESS_NOT_FOUND
+  }
+
   private val trackedStoriesProgress: Map<String, TrackedStoryProgress> by lazy { createInitialStoryProgressState() }
 
+  private val cacheStoreMap = mutableMapOf<ProfileId, PersistentCacheStore<TopicProgressDatabase>>()
+
+  private fun retrieveCacheStore(profileId: ProfileId): PersistentCacheStore<TopicProgressDatabase> {
+    return if (profileId in cacheStoreMap) {
+      cacheStoreMap[profileId]!!
+    } else {
+      val cacheStore =
+        cacheStoreFactory.createPerProfile(CACHE_NAME, TopicProgressDatabase.getDefaultInstance(), profileId)
+      cacheStoreMap[profileId] = cacheStore
+      cacheStore
+    }
+  }
+
+  fun retrieveStoryProgress(storyId: String): StoryProgress {
+    return createStoryProgressSnapshot(storyId)
+  }
+
   /**
-   * Records the specified chapter completed within the context of the specified story. Returns a [LiveData] that
+   * Records the specified chapter completed within the context of the specified exploration, story, topic. Returns a [LiveData] that
    * provides exactly one [AsyncResult] to indicate whether this operation has succeeded. This method will never return
    * a pending result.
    */
-  fun recordCompletedChapter(storyId: String, explorationId: String): LiveData<AsyncResult<Nothing?>> {
-    return try {
-      trackCompletedChapter(storyId, explorationId)
-      MutableLiveData(AsyncResult.success(null))
-    } catch (e: Exception) {
-      MutableLiveData(AsyncResult.failed(e))
+  fun recordCompletedChapter(
+    internalProfileId: Int,
+    topicId: String,
+    storyId: String,
+    explorationId: String
+  ): LiveData<AsyncResult<Any?>> {
+    val profileId = ProfileId.newBuilder().setInternalId(internalProfileId).build()
+    val deferred = retrieveCacheStore(profileId).storeDataWithCustomChannelAsync(updateInMemoryCache = true) {
+      // Replace this with checking if the id pair already exists.
+      val chapterPlayState = ChapterPlayState.COMPLETED
+      val storyProgressNew = StoryProgress.newBuilder().putChapterProgress(explorationId, chapterPlayState).build()
+      val topicProgress = TopicProgress.newBuilder().putStoryProgress(storyId, storyProgressNew).build()
```

Comment:
Is this overwriting any other stories that have had progress made in this topic? Similarly, is the story progress newBuilder() line above also overwriting any other chapters that were completed? Are there tests to verify that multiple chapters can be marked as completed for the same & different stories, as well as different topics?

Reply:
Actually above reply is incorrect as after investigating by implementing the app module functions, the progress was indeed getting overwritten. So updated the code accordingly and added test cases.